src/platform/riot/uart_channel.c

```diff
@@ -0,0 +1,163 @@
+
+static lf_ret_t UARTPollChannel_open_connection(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Open connection");
+  (void)untyped_self;
+  return LF_OK;
+}
+
+static void UARTPollChannel_close_connection(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Close connection");
+  (void)untyped_self;
+}
+
+static void UARTPollChannel_free(NetworkChannel *untyped_self) {
+  UART_CHANNEL_DEBUG("Free");
+  (void)untyped_self;
+}
+
+static bool UARTPollChannel_is_connected(NetworkChannel *untyped_self) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+  return self->state == NETWORK_CHANNEL_STATE_CONNECTED;
+}
+
+static lf_ret_t UARTPollChannel_send_blocking(NetworkChannel *untyped_self, const FederateMessage *message) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+
+  if (self->state == NETWORK_CHANNEL_STATE_CONNECTED) {
+    int message_size = serialize_to_protobuf(message, self->write_buffer, UART_CHANNEL_BUFFERSIZE);
+
+    UART_CHANNEL_DEBUG("Sending Message of Size: %i", message_size);
+
+    uart_write(self->uart_dev, self->write_buffer, message_size);
+    return LF_OK;
+  } else {
+    return LF_ERR;
+  }
+}
+
+static void UARTPollChannel_register_receive_callback(NetworkChannel *untyped_self,
+                                                      void (*receive_callback)(FederatedConnectionBundle *conn,
+                                                                               const FederateMessage *msg),
+                                                      FederatedConnectionBundle *conn) {
+  UART_CHANNEL_INFO("Register receive callback");
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+
+  self->receive_callback = receive_callback;
+  self->federated_connection = conn;
+}
+
+void _UARTPollChannel_interrupt_callback(void *arg, uint8_t received_byte) {
+  UARTPollChannel *self = (UARTPollChannel *)arg;
+  const uint32_t minimum_message_size = 12;
+
+  self->receive_buffer[self->receive_buffer_index] = received_byte;
+  self->receive_buffer_index++;
+
+  if (self->receive_buffer_index >= minimum_message_size) {
+    if (self->super.super.mode == NETWORK_CHANNEL_MODE_ASYNC) {
+      cond_signal(&((UARTAsyncChannel *)self)->receive_cv);
+    }
+  }
+}
+
+void UARTPollChannel_poll(NetworkChannel *untyped_self) {
+  UARTPollChannel *self = (UARTPollChannel *)untyped_self;
+  const uint32_t minimum_message_size = 12;
+
+  while (self->receive_buffer_index > minimum_message_size) {
+    int bytes_left = deserialize_from_protobuf(&self->output, self->receive_buffer, self->receive_buffer_index);
+    UART_CHANNEL_DEBUG("Bytes Left after attempted to deserialize %d", bytes_left);
+
+    if (bytes_left >= 0) {
+      int receive_buffer_index = self->receive_buffer_index;
+      self->receive_buffer_index = bytes_left;
+      memcpy(self->receive_buffer, self->receive_buffer + (receive_buffer_index - bytes_left), bytes_left);
```

Comment:
For this to work, `UartPollChannel_poll` must always be called within a critical section, so that interrupts are disabled. A few other places I added `_locked` postfix to functions that has to be called from critical sections. Maybe this would be wise for this one as well?

Reply:
I would suggest we protect updates to th `receive_buffer_index` but the `receive_buffer` I would not protect.

Reply:
Hmm, I do not think this is safe. What happens e.g. if the interrupt occurs between line 79 and 84? Say we have parsed out a message, and that there are 4 bytes left. Well, if the interrupt occurs, then really there should be 5 bytes left, so when we do the memcpy to copy the bytes back to the beginning of the buffer, we forget the fifth byte and it is lost.

I think the memcpy has to be part of the critical section as well

Reply:
What I would prefer to this is turning the `receive_buffer_index` into an atomic, then we could get rid of that lock.

I think only protecting the `receive_buffer_index` is enough, because the interrupts are sequential so the incrementing is sequential and overwriting the `receive_buffer_index` also gets sequenialized because we lock this region.